### DISCUSSION TITLE
Mod Configuration API

### DISCUSSION THREAD

**Discussion by zkxs** [neos-modding-group/NeosModLoader#4] [Ideas]
## Use cases
- mod devs want to save/load configs
- GUI dev wants to be able to dynamically build a config UI
- modloader itself should be kept as minimal and simple as possible

## Concepts
- a config will be a named, ordered collection of config items
  - naming convention should be documented
  - configs will be defined by mods as the mods are loaded
- config items are a single value of an allowed type
  - why limit types? To make things easier for the GUI devs.
  - allowed types
    - bool
    - string
    - int
    - float
    - array of any of the above types?
    - do we need more types than this?
    - config items will have metadata
      - externally accessible: bool. If true, indicates that this config item is allowed to be changed from outside of the creating mod

## Requirements
- load a config
- save a config
- config versioning?
  - config definitions require defaults
  - unrecognized configs will be droped
- config metadata
  - type
  - min/max/increment for floats/ints?
  - hidden vs visible in UI?
- config file is NOT designed to be user-editable. Comments and ordering will be clobbered.

**Comment by EIA485** [comment]
there should be a color type and some way to handle enums

**Reply by zkxs** [reply]
My current plan is to just ditch the whole concept of allowed types entirely and allow everything. We'll see if it works out.
